src/Nethermind/Nethermind.State/VerkleStateTree.cs

```diff
@@ -77,6 +78,61 @@ public void SetStorage(StorageCell cell, byte[] value)
         Insert(key, value);
     }
 
+    public void BulkSet(IDictionary<StorageCell, byte[]> values)
+    {
+        // Put the sets into a list to be sorted
+        using ArrayPoolList<KeyValuePair<StorageCell, byte[]>> theList = new ArrayPoolList<KeyValuePair<StorageCell, byte[]>>(values.Count, values);
+
+        // Sort by address and index.
+        theList.AsSpan().Sort((kv1, kv2) =>
+        {
+            int addressCompare = kv1.Key.Address.CompareTo(kv2.Key.Address);
+            return addressCompare != 0 ? addressCompare : kv1.Key.Index.CompareTo(kv2.Key.Index);
+        });
+
+        ActionBlock<(Hash256, Dictionary<byte, byte[]>)> insertStem =
+            new ActionBlock<(Hash256, Dictionary<byte, byte[]>)>((item) =>
+            {
+                InsertStemBatch(item.Item1.Bytes[..31], item.Item2.Select(kv => (kv.Key, kv.Value)));
+            },
+            new ExecutionDataflowBlockOptions()
+            {
+                MaxDegreeOfParallelism = Environment.ProcessorCount,
+            });
+
+        Hash256 currentStem = Hash256.Zero;
+        Dictionary<byte, byte[]> stemValues = new Dictionary<byte, byte[]>();
+        foreach (KeyValuePair<StorageCell, byte[]> kv in theList)
+        {
+            // Because of the way the mapping works
+            Hash256 theKey = AccountHeader.GetTreeKeyForStorageSlot(kv.Key.Address.Bytes, kv.Key.Index);
+
+            if (!currentStem.Bytes[..31].SequenceEqual(theKey.Bytes[..31]))
+            {
+                // Different stem, will attempt to insert the stem batch
+                if (stemValues.Count != 0)
+                {
+                    // Stem is different and stemValues have value.
+                    insertStem.Post((currentStem, stemValues));
+                    stemValues = new Dictionary<byte, byte[]>();
+                }
+
+                // And set the next stem
+                currentStem = theKey;
+            }
+
+            stemValues[theKey.Bytes[31]] = kv.Value;
+        }
+
+        if (stemValues.Count != 0)
+        {
+            insertStem.Post((currentStem, stemValues));
+        }
+
+        insertStem.Complete();
+        insertStem.Completion.Wait();
+    }
```

Comment:
this specific optimization 
do we know how much performance gain are we getting just because of this?

Reply:
the reason for this question - we can do this kind of optimiztion for even larger datasete
right now when we insert - we calculate some commitment that is grouped by stems
but we can skip that and do all the calculations when call Commit()

Reply:
About 50% overall on balance test. About 10% on address only. Yes, I have the exact same thought on the Commit part.

Reply:
The 50% is only because of the parallelism.